### PR TITLE
Do not cache packages in images

### PR DESCRIPTION
**Pull request by eapolinario**
# TL;DR
Disable caching of packages in images

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Pass the `--no-cache-dir` flag in invocation of `pip install` in dockerfiles. This will have a bigger impact in the agent image, since currently we are using about ~800MB:
```
 docker run --rm -it ghcr.io/flyteorg/flyteagent:1.10.0 du -kh --max-depth=1 /root/.cache
813M    /root/.cache/pip
813M    /root/.cache
``` 

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Review by pingsutw — APPROVED** [review]

File: Dockerfile [hunk 1]
```diff
@@ -13,10 +13,11 @@ ARG DOCKER_IMAGE
 RUN apt-get update && apt-get install build-essential -y
 
 # Pod tasks should be exposed in the default image
-RUN pip install -U flytekit==$VERSION \
+RUN pip install --no-cache-dir -U flytekit==$VERSION \
 	flytekitplugins-pod==$VERSION \
 	flytekitplugins-deck-standard==$VERSION \
-	scikit-learn
+	scikit-learn \
+	&& :
```
Comment (jeevb) [inline Dockerfile:20]: Why is this necessary?

**Review by jeevb — APPROVED** [review]

File: Dockerfile (hunk 1 above)
Comment (kumare3) [reply Dockerfile:20]: Great catch should not be needed

File: Dockerfile.agent [hunk 2]
```diff
@@ -8,7 +8,7 @@ ARG VERSION
 RUN apt-get update && apt-get install build-essential -y
 
 RUN pip install prometheus-client
```
Comment (jeevb) [inline Dockerfile.agent:10]: Missed one. Why is this separate anyway?

**Comment by jeevb** [comment]
build-essential adds an insignificant overhead. It should only be needed for building wheels, and not needed in the final image. Also apt cache is not removed.

File: Dockerfile (hunk 1 above)
Comment (eapolinario) [reply Dockerfile:20]: Adding `&& :` is a no-op. This helps in maintaining git history (if we want to remove `scikit-learn` or install another package you don't have to add another symbol to that line).

Comment (jeevb) [reply Dockerfile:20]: Ah cool!

File: Dockerfile.agent (hunk 2 above)
Comment (pingsutw) [reply Dockerfile.agent:10]: since we don't want to add `prometheus-client` to the default flytekit dependencies.

**Comment by eapolinario** [comment]
> build-essential adds an insignificant overhead. It should only be needed for building wheels, and not needed in the final image. Also apt cache is not removed. We can clean this up in a subsequent PR.

we need `build-essential` for this exact reason. We build wheels for some of our dependencies, e.g. `pyspark` in the case of the agents. I also did a quick test removing the apt cache and we could shave another 20-30MB.
